File: src/BaekJoon/java/baekjoon_2178.java
package BaekJoon.java;
// https://www.acmicpc.net/problem/2178

import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class baekjoon_2178 {
	static class coordinate {
		int x;
		int y;
		int count;

		coordinate() {
			x = 1;
			y = 1;
			count = 1;
		}

		coordinate(int x, int y, int count) {
			this.x = x;
			this.y = y;
			this.count = count;
		}
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);

		int n = sc.nextInt();
		int m = sc.nextInt();
		int[][] field = new int[n+2][m+2];
		boolean[][] c = new boolean[n+2][m+2];

		sc.nextLine();

		for(int i = 1; i < field.length-1; i++) {
			String s = sc.nextLine();

			for(int j = 0; j < s.length(); j++)
				field[i][j+1] = s.charAt(j) - 48;
		}

		bfs(field, c);

		sc.close();
	}

	public static void bfs(int[][] field, boolean[][] c) {
		Queue<coordinate> q = new LinkedList<coordinate>();
		coordinate coor = null;

		c[1][1] = true;
		q.add(new coordinate());

		while(!q.isEmpty()) {
			coor = q.poll();

			if(coor.x == field.length-2 && coor.y == field[1].length-2)
				break;

			// check can move up
			if(field[coor.x-1][coor.y] == 1 && !c[coor.x-1][coor.y]) {
				c[coor.x-1][coor.y] = true;
				q.add(new coordinate(coor.x-1, coor.y, coor.count+1));
			}

			// check can move right
			if(field[coor.x][coor.y+1] == 1 && !c[coor.x][coor.y+1]) {
				c[coor.x][coor.y+1] = true;
				q.add(new coordinate(coor.x, coor.y+1, coor.count+1));
			}

			// check can move down
			if(field[coor.x+1][coor.y] == 1 && !c[coor.x+1][coor.y]) {
				c[coor.x+1][coor.y] = true;
				q.add(new coordinate(coor.x+1, coor.y, coor.count+1));
			}

			// check can move left
			if(field[coor.x][coor.y-1] == 1 && !c[coor.x][coor.y-1]) {
				c[coor.x][coor.y-1] = true;
				q.add(new coordinate(coor.x, coor.y-1, coor.count+1));
			}


		} // end of while

		System.out.println(coor.count);
	}
}

File: src/BaekJoon/python/baekjoon_2178.py
# https://www.acmicpc.net/problem/2178

import sys
readLine = lambda : sys.stdin.readline().strip()
n, m = map(int, readLine().split())

field = [[int(row) for row in readLine()] for col in range(n)]

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def bfs() :
	field[0][0] = 2
	q = []
	q.append((0, 0, 1))

	while(len(q)) :
		coor = q.pop(0)

		if n-1 == coor[0] and m-1 == coor[1] :
			print(coor[2])
			break

		for i in range(4) :
			currX, currY, count = coor
			currX += dx[i]
			currY += dy[i]

			if (currX >= 0 and currY >= 0 and currX < n and currY < m) and (field[currX][currY] is 1) :
				field[currX][currY] = 2
				count += 1
				q.append((currX, currY, count))

bfs()